src/headers.jl

```diff
@@ -0,0 +1,246 @@
+# See https://httpwg.github.io/specs/rfc7230.html#rule.token.separators
+const validHeaderBytes = IntSet(Int(c) for c in
+   raw"!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUWVXYZ^_`abcdefghijklmnopqrstuvwxyz|~"
+)
+
+validHeaderChar(c::Char) = isascii(c) && UInt8(c) in validHeaderBytes
+
+function canonicalHeaderKey(s::String)
+    # fast path for valid header keys
+    upper = true
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+        if upper && 'a' <= c <= 'z'
+            return _chk(s)
+        end
+        if !upper && 'A' <= c <= 'Z'
+            return _chk(s)
+        end
+
+        upper = c == '-'
+    end
+
+    return s
+end
+
+const toUpper = 'A' - 'a'
+
+function _chk(s::String)
+    # confirm its fixable before changing anything
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+    end
+
+    a = copy(Array{UInt8}(s))
+    upper = true
+    for (i, b) in enumerate(a)
+        if upper && UInt8('a') <= b <= UInt8('z')
+            b += toUpper
+        elseif !upper && UInt8('A') <= b <= UInt8('Z')
+            b -= toUpper
+        end
+        @inbounds a[i] = b
+        upper = b == UInt8('-')
+    end
+
+    return String(a)
+end
+
+"""
+Internal type used to enable fast merge, copy, and filter of Headers
+"""
+immutable AlreadyCanonical
```

Comment:
What do you think about getting rid of `AlreadyCanonical` and just having a boolean flag to the `Headers` constructor? It seems a little less heavy than needing another type.